php_phongo.c

```diff
@@ -451,10 +451,65 @@ mongoc_bulk_operation_t *phongo_bulkwrite_init(zend_bool ordered) { /* {{{ */
 #define PHONGO_WRITECONCERN_ALLOWED   0x01
 #define PHONGO_READPREFERENCE_ALLOWED 0x02
 
-static int process_read_preference(zval *option, zval **zreadPreference TSRMLS_DC)
+static bool process_read_concern(zval *option, bson_t *mongoc_opts TSRMLS_DC)
+{
+	if (Z_TYPE_P(option) == IS_OBJECT && instanceof_function(Z_OBJCE_P(option), php_phongo_readconcern_ce TSRMLS_CC)) {
+		const mongoc_read_concern_t *read_concern = phongo_read_concern_from_zval(option TSRMLS_CC);
+
+		if (!mongoc_read_concern_append((mongoc_read_concern_t*)read_concern, mongoc_opts)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error appending \"%s\" option", "ReadPreference");
+		}
+	} else {
+		phongo_throw_exception(
+			PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC,
+			"Expected 'readConcern' option to be 'MongoDB\\Driver\\ReadPreference', %s given",
+			zend_get_type_by_const(Z_TYPE_P(option))
+		);
+		return false;
+	}
+	return true;
+}
+
+static int phongo_do_select_server(mongoc_client_t *client, bson_t *opts, zval *zreadPreference, int server_id TSRMLS_DC)
+{
+	bson_error_t error;
+	uint32_t selected_server_id;
+
+	if (server_id > 0) {
+		bson_append_int32(opts, "serverId", -1, server_id);
+		selected_server_id = server_id;
+	} else {
+		mongoc_server_description_t  *selected_server = NULL;
+
+		selected_server = mongoc_client_select_server(client, false, (zreadPreference ? phongo_read_preference_from_zval(zreadPreference TSRMLS_CC) : mongoc_client_get_read_prefs(client)), &error);
+		if (selected_server) {
+			selected_server_id = mongoc_server_description_id(selected_server);
+			bson_append_int32(opts, "serverId", -1, selected_server_id);
+			mongoc_server_description_destroy(selected_server);
+		} else {
+			/* Check for connection related exceptions */
+			if (!EG(exception)) {
+				phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
+			}
+
+			return false;
```

Comment:
I suppose this should be zero to jive with the return type. Alternatively, initialize `selected_server_id` to zero up top and remove this line entirely.

Reply:
done